### PR TITLE
PL: Test PyTorch 2.0 pre-release on CPU and CUDA

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Adds CI for PT 2.0 pre release for CPU and GPU

A follow-up will do the same for Fabric

### CODE REVIEW

File: .github/workflows/ci-tests-pytorch.yml
Comment (akihironitta) [inline .github/workflows/ci-tests-pytorch.yml:64]:
```suggestion
          # pytorch 2.0 Tests (currently pre-release, change this once it's released)
```

**Review by akihironitta — APPROVED** [review]

File: .github/workflows/ci-tests-pytorch.yml
Comment (Borda) [inline .github/workflows/ci-tests-pytorch.yml:94]:
```suggestion
      if: ${{ matrix.requires != 'oldest' && matrix.release != 'pre' }}
```

File: dockers/base-cuda/Dockerfile
Comment (Borda) [inline dockers/base-cuda/Dockerfile:24]: I would not include pre option to dockers, lets keep it simple :)

Comment (justusschock) [reply dockers/base-cuda/Dockerfile:24]: This is still left from when I wanted to follow another path for GPU CI.

File: requirements/pytorch/adjust-versions.py
Comment (carmocca) [inline requirements/pytorch/adjust-versions.py:53]: Can you explain why the changes to this file are necessary?

Can't we just skip the step where they are initially fixed, instead of removing them again? I feel like this is a difficult solution in the code to what would be simple in CI

File: .github/workflows/ci-tests-pytorch.yml
Comment (carmocca) [inline .github/workflows/ci-tests-pytorch.yml:67]: These should be added to the checkgroup config

File: .azure/gpu-tests-pytorch.yml
Comment (carmocca) [inline .azure/gpu-tests-pytorch.yml:58]: Why not create a 2.0 docker image? We'll need it later anyways

Comment (Borda) [reply .azure/gpu-tests-pytorch.yml:58]: That sounds good to me =)

Comment (justusschock) [reply .azure/gpu-tests-pytorch.yml:58]: I wanted to but @Borda didn't like changes to the dockerfile

File: .github/workflows/ci-tests-pytorch.yml
Comment (justusschock) [reply .github/workflows/ci-tests-pytorch.yml:67]: shouldn't we wait with this until they have an official rc?

File: requirements/pytorch/adjust-versions.py
Comment (justusschock) [reply requirements/pytorch/adjust-versions.py:53]: This is necessary since a won't be selected if there is any pinned version (even with `--pre`). So I still had to remove any version specifiers.

Comment (justusschock) [reply requirements/pytorch/adjust-versions.py:53]: I'll revert this once I have a working docker image

File: .azure/gpu-tests-pytorch.yml
Comment (Borda) [reply .azure/gpu-tests-pytorch.yml:58]: True, I did revert it, but maybe it would be good call; sorry for moving there and back...

Comment (Borda) [reply .azure/gpu-tests-pytorch.yml:58]: as it is solution just for next a few weeks I would stick with the simple way and don't build special images as that would also mean additional adjustments on workflow side so the RC are not overwritten due to `FREEZE_REQUIREMENTS`

**Review by Borda — APPROVED** [review]
LGTM from the CI stand of view :tada:

File: .github/workflows/ci-tests-pytorch.yml
Comment (justusschock) [inline .github/workflows/ci-tests-pytorch.yml:110]:
```suggestion
```

**Review by awaelchli — APPROVED** [review]

File: dockers/base-cuda/Dockerfile
Comment (carmocca) [inline dockers/base-cuda/Dockerfile:107]: Where did the strategy install go?

File: dockers/base-cuda/Dockerfile
Comment (carmocca) [inline dockers/base-cuda/Dockerfile:100]: This is unused?

```suggestion
    pip install -r requirements/pytorch/base.txt --no-cache-dir --find-links ${TORCH_URL} && \
```

File: dockers/base-cuda/Dockerfile
Comment (carmocca) [inline dockers/base-cuda/Dockerfile:96]: Dockers use 2 spaces

```suggestion
      python ./requirements/pytorch/adjust-versions.py $fpath ${PYTORCH_VERSION} ; \
```

File: .github/workflows/ci-tests-pytorch.yml
Comment (carmocca) [reply .github/workflows/ci-tests-pytorch.yml:67]: It's out now

File: .azure/gpu-tests-pytorch.yml
Comment (carmocca) [inline .azure/gpu-tests-pytorch.yml:112]: Unused?

```suggestion
    - bash: pip install -e .[extra,test,examples] -U --find-links ${TORCH_URL}
```

**Comment by carmocca** [comment]
Looks like `onnx` is now a required dependency for the onnx tests.

File: .github/workflows/ci-tests-pytorch.yml
Comment (justusschock) [reply .github/workflows/ci-tests-pytorch.yml:67]: yes, as @Borda shared on slack, we would merge it without adding it here, to first make sure they pass and once we're confident it works as it should, we would add it to checkgroup.
This PR only adds CI but does not aim at fixing issues.

Comment (carmocca) [reply .github/workflows/ci-tests-pytorch.yml:67]: Please don't merge this with failing jobs, even if they are unmarked as required. CI should always be green.

I'm fine with fixing the issues separately, but then I would fix those first. You can work here to validate the changes and pull them to a separate PR once they are passing

File: dockers/base-cuda/Dockerfile
Comment (Borda) [reply dockers/base-cuda/Dockerfile:107]:
```suggestion
    pip install -r requirements/pytorch/devel.txt -r requirements/pytorch/strategies.txt --no-cache-dir --find-links ${TORCH_URL}
```

File: .azure/gpu-tests-pytorch.yml
Comment (Borda) [reply .azure/gpu-tests-pytorch.yml:112]: can be dropped

File: .github/workflows/ci-tests-pytorch.yml
Comment (Borda) [reply .github/workflows/ci-tests-pytorch.yml:67]: > Please don't merge this with failing jobs, even if they are unmarked as required. CI should always be green.

ls help us to make it green :)

File: requirements/pytorch/base.txt
Comment (carmocca) [inline requirements/pytorch/base.txt:13]: Why not keep this in the test requirements?

**Comment by Borda** [comment]
seems some size leaking:
```
>       assert cuda_memory < model.start_cuda_memory
E       assert 17039872 < 2048
E        +  where 2048 = BoringModelGPU(\n  (layer): Linear(in_features=32, out_features=2, bias=True)\n).start_cuda_memory
```

File: requirements/pytorch/base.txt
Comment (Borda) [reply requirements/pytorch/base.txt:13]: becase it is needed for PT 2.0, they have it as optional package so it crash with error you need to install it :)

Comment (Borda) [reply requirements/pytorch/base.txt:13]: and in tets we still keep runitme version... :chipmunk:

**Review by carmocca — CHANGES_REQUESTED** [review]
Please don't merge this until 2.0 failures are resolved

File: requirements/pytorch/base.txt
Comment (carmocca) [reply requirements/pytorch/base.txt:13]: I don't think that's right. The onnx installation should still be optional: https://github.com/Lightning-AI/lightning/pull/16891

Comment (Borda) [reply requirements/pytorch/base.txt:13]: ok, lets merge #16764 here instead of master :otter:

File: tests/tests_pytorch/strategies/test_ddp.py
Comment (carmocca) [inline tests/tests_pytorch/strategies/test_ddp.py:71]: PyTorch 2.0 changed the backend validation logic, and this is now hanging instead of raising the ValueError

Either way, we can remove this test because the exception is raised inside Torch, so we don't need to test that

**Review by carmocca — APPROVED** [review]
All green now

**Pull request merged**
